### ISSUE TITLE
About the paper " LLMS ARE SINGLE-THREADED REASONERS: DEMYS TIFYING THE WORKING MECHANISM OF SOFT THINK ING"

### ISSUE THREAD

**Issue by albert-jin** [eric-ai-lab/Soft-Thinking#16]
<img width="1054" height="786" alt="Image" src="https://github.com/user-attachments/assets/0c76a62a-1259-485f-b7f3-3caed10c9412" />
In this paper, the authors said they reproduce your code, and the results of your "soft-thinking" is shown in this picture's table.

however, your report results in your paper is  MATH 500 is 98.00， but his paper is reported with 91.60, all under "Deepseek-R1-Distill-Qwen-32B",

So I have thought a question and such a misunderstanding, (Based on his paper, this used your code to reproduce the results. And the README file of them said they also used you .py file named "run_sglang_softthinking.py", So I think they must used your LLM judging's logics.)

As you said in your previous closed issue, the LLM judge is important to determine whether LLM reasoning is correct. So the only reason I will guess is that： “their code did not use the same LLM Judge logics with yours.”

I wish you can check their paper, and code: "https://github.com/WJMacro/Randomized-Soft-Thinking/tree/main" However, their code is not updated, they only update the content of their README.md.

**Comment by albert-jin** [comment]
<img width="707" height="819" alt="Image" src="https://github.com/user-attachments/assets/856dcc5c-daca-4dcf-8cbf-97645507c9d1" />

<img width="700" height="299" alt="Image" src="https://github.com/user-attachments/assets/9f64781e-5993-498d-9563-b73f536f176e" />

**Comment by namezhenzhang** [comment]
Soft Thinking is not very stable while generating, so hyper-parameter search, Cold Stop, and LLM judge are necessary for evaluation. Otherwise, we can not extract the answer with only rule-based scripts.

**Comment by albert-jin** [comment]
> Soft Thinking is not very stable while generating, so hyper-parameter search, Cold Stop, and LLM judge are necessary for evaluation. Otherwise, we can not extract the answer with only rule-based scripts.

Yes, the final aiming/target for validating whether the reasoning is correct is not about how to match the ground truth answer in the reasoning chain. It is important on whether the overall reasoning is right and whether the LLM can reason the final answer.

我很赞同judge的逻辑，因为LLM推理本身这个任务应该关注放在推理链和推理是否正确上，LLM judge用来匹配原来用作推理的LLM是否能得到答案，这个只是辅助工具，不应该恶意去掉。

先不管那么多了，等他们这篇randomized soft-thinking公开后，我们再好好的看一看他们究竟魔改了你的什么，导致最终结果这么夸张离谱，哈哈哈

**Comment by albert-jin** [comment]
发现了一个问题，Wu的paper，里面引入了gumbel softmax，为的目的就是让随机采样的过程能够反向传播，但问题是，您的soft-thinking既然要吸纳他们的两个idea，迪利克雷 & gumbel, 代码可能不全。

具体来说，其中的迪利克雷 我看了，没啥大问题，也就那么几行。但是：关于gumbel, 我有点问题：一方面，他代码让你pull-request的代码中，没有任何RL的逻辑，但他们文章中写了需要反向传播优化，需要强化学习，。但我找你的diff文件中，没有找到任何跟gumbel的反向传播有关的内容，这和他们说引用gumbel的故事不太一样。

难道Wu等人的这篇paper，里面真的就只只只是用了gumbel的前向传播吗？

**Comment by albert-jin** [comment]
总结：我认为Wu他们文章的，在开头前面，批判soft-thinking的动机很好，但是他在提出他自己的gumbel trick的动机，不是很清晰

**Comment by albert-jin** [comment]
> 可能原文中的“我们使用了Soft Thinking(Zhang, et al)的SGL实现”容易引起误解，我们会考虑增加额外注释表明基线并非直接对Soft Thinking(Zhang, et al)结果的复现，实际上等价于Soft Thinking w/o Cold Stop。

这样就很好了，给Wu点个赞，你的paper，需要明确一下，你的故事线不应该单单描述soft-thinking原作者的单线程的缺点，因为基本上所有的followers，包括我，都知道，实际上起作用的是“你合理的解决了冷停的问题”，是你新引入的gumbel或迪利克雷，让token的选择，从soft的加权混合，变成了离散，从而顺路解决了冷停&无限序列生成的问题。结论：我个人建议你需要在Meta实验里面，补充关于冷停，也是你的一大动机（只是个人意见）。

**Comment by albert-jin** [comment]
> 关于Evaluation中Judge的使用，在我们的所有实验中都是使用本地的Math Verify进行。在同等对比条件下，不使用LLM as a Judge并不影响内部比较的公平性，算不上恶意取消。

和你的想法一样，是否使用LLM judge 来辅助解决Math judge，这些类似的问题，本质属于答案匹配相关的，跟我们本身的任务无关，当然，借助LLM judge 判断出正确的答案，当然能证明LLM reason的推理也是没问题的，我很赞同你的说法：用什么样的judge本身没什么，只要judge的逻辑言之有理即可，能保证尽量准确的反应和检测出LLM的推理是否正确即可！ 很赞同Wu的这部分说法

**Comment by albert-jin** [comment]
> 但fork后发现soft thinking作者已经集成了我们提供的采样方法，

明白，所以soft并非集成你paper的所有idea，尤其是重点RL方面没有，先不管soft有没有完全集成你的，你先好好跑RL吧，加油

**Comment by albert-jin** [comment]
> 而关于您的第二个问题，即引入Gumbel Softmax的动机，我们也有充分合适的理由。
> 
> 首先你对Gumbel随机变量存在一些误解。在传统机器学习中面临argmax操作不可导时，大家会常说可以用gumbel max来代替。但这里所提的gumbel max其实是gumbel max+straight through estimation两个技术的复合。前者利用Gumbel随机变量符合Luce选择公理的性质实现随机选择，后者则保证了梯度可以通过不可导的选择操作进行前传。在我们面临的问题中，我们希望利用可控的随机性解决vanilla soft thinking的贪心缺陷。此时，我们只使用了Gumbel随机变量的选择性质，通过在logits上引入Gumbel变量，我们实现了对合理token的重排序。而Softmax可以进一步将Gumbel logits归一化为合法的概率向量用于输入。这一过程是纯粹的解码后采样操作，故而不需要直通估计。事实上Gumbel本人在1935年提出这种极值分布时，直通估计（2013）尚不存在。所以所谓“gumbel 的核心优势是为了能够采样过程能够反向传播而设计的”更无从谈起了。
> 
> 同时，我们的论文从Luce选择公理角度说明了，Gumbel Softmax得到的分布中token的prob是从原模型概率分布中进行多次不放回抽样的平滑结果。这使得这种采样方式自然的成为token sampling(一次不放回抽样)的多token扩展。只使用gumbel的分布性质而不使用直通估计也是完全合理的技术，例如Stochastic Beams(Kool, et al)一文也采用了类似的采样技巧。
> 
> 关于强化学习部分，我们在附录中展示了引入随机采样的前景，即可以实现Test Time Scaling和RL。其中RL只需要计算policy ratio。这可以通过Category Distribution或者重参数化技巧来实现。同样也不需要利用直通估计技巧。
=============
这个我就不回复了，看不懂直接略过，但我明白你的大概意思：（文中说过）通过合理的gumbel噪音，让LLM不只是贪心，而是随机选择，让LLM不得不多探索不同的推理路径~ 【有一丢丢像VAE，通过加噪音提高模型的泛化性，哈哈哈，也有点像dropout，通过对网络层加噪音，提升网路中每个神经元的泛化性】

**Comment by albert-jin** [comment]
我们没有遇到这个问题

**Comment by liushengjing-cmd** [comment]
@albert-jin 我想请问一下有没有遇见过，2卡和4卡跑的时候，指标差距能达到3-4个点的情况，目前我发现复现baseline的时候2卡L40跑的结果接近baseline，4卡则差距3-4个点。
